### PR TITLE
DT: Add DT_DMA macros to help usage in dma clients drivers

### PR DESCRIPTION
**Pull request by erwango**
Add DT_DMA macros to be used in dma clients
Add matching tests.
Also update stm32 i2s driver to use them.

EDIT: 
Docs updated
Added accessor macro (I don't plan to add more)
Miss tests on additional macros

### CODE REVIEW

**Comment by erwango** [comment]
@galak I think what is missing is :
- _INST macros
- macros to access "channels" by index (on top of "_BY_NAME" already provided)

Though I don't plan to provide macros to access cells by index cells need names and always have so far. Besides we already need IDX accessors for channels, so I want to avoid IDX_BY_IDX stuff.

File: dts/bindings/dma/dma.yaml [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 properties:
     dmas:
-      required: true
+      required: false
```
Comment (galak) [inline dts/bindings/dma/dma.yaml:9]: what's the thinking on marking this as `required: false`?

Comment (erwango) [reply dts/bindings/dma/dma.yaml:9]: dma.yaml needs to be included in clients binding.
If this is marked true, all clients need to add dmas props, while this should remain optional

Comment (galak) [reply dts/bindings/dma/dma.yaml:9]: We should than just move this into base.yaml.  I assume its ok for this to be optional for st,stm32-i2s.yaml

File: include/devicetree/dma.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * @defgroup devicetree-dmas Devicetree dma API
+ * @{
+ */
+
+/**
+ * @brief Get dma controller "name" (label property) by name
+ *
+ * It's an error if the dma controller referenced by the phandle
+ * in property "dmas" at index "idx" has no label property.
```
Comment (mbolivar-nordic) [inline include/devicetree/dma.h:28]: "idx" doesn't seem to be an argument to this macro. Are there some copy/paste issues in this docstring?

File: scripts/dts/gen_defines.py [hunk 3]
```diff
@@ -488,6 +488,8 @@ def controller_and_data_macros(entry, i, macro):
     ret[f"{macro}_IDX_{i}_NAME"] = quote_str(entry.name)
     # DT_N_<node-id>_P_<prop-id>_NAME_<NAME>_PH
     ret[f"{macro}_NAME_{name}_PH"] = f"DT_{entry.controller.z_path_id}"
+    # DT_N_<node-id>_P_<prop-id>_NAME_<NAME>_EXISTS
```
Comment (mbolivar-nordic) [inline scripts/dts/gen_defines.py:491]: Could you please update macros.bnf in the docs too?

File: dts/bindings/dma/dma.yaml (hunk 1 above)
Comment (erwango) [reply dts/bindings/dma/dma.yaml:9]: Moved to base.yaml.
Though mark them as required in stm32 i2s.

File: include/devicetree/dma.h (hunk 2 above)
Comment (erwango) [reply include/devicetree/dma.h:28]: Yes, there are. I haven't paid much attention to the doc part yet, wanted to get initial feedbacks first.
Will update.

File: scripts/dts/gen_defines.py (hunk 3 above)
Comment (erwango) [reply scripts/dts/gen_defines.py:491]: Totally forgotten this doc. Will update as well. Thanks

Comment (erwango) [reply scripts/dts/gen_defines.py:491]: Done, Please suggest is not correct.

Comment (mbolivar-nordic) [reply scripts/dts/gen_defines.py:491]: This part looks fine. I see now that it's covered by the existing `prop-suf`:

```
prop-suf = 1*( "_" gen-name ["_" dt-name] )
```

In this case, `prop-suf` has to match `_NAME_<NAME>_EXISTS`, which it does, basically as a gen-name + dt-name + gen-name sequence.

File: dts/bindings/base/base.yaml [hunk 4]
```diff
@@ -70,3 +70,13 @@ properties:
...
+
+    dmas:
+      required: false
+      type: phandle-array
+      description: DMA channels specifiers
+
+    dma-names:
+      required: false
+      type: string-array
+      description: Probvided names of DMA channel specifiers
```
Comment (mbolivar-nordic) [inline dts/bindings/base/base.yaml:82]:
```suggestion
      description: Provided names of DMA channel specifiers
```

File: include/devicetree/dma.h [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+ * @param name reference of the tatgeted dma channel specifier
+ * @return the value of "cell" inside the specifier
+ */
+#define DT_INST_DMAS_CELLS_BY_NAME(inst, cell_name, name) \
+	DT_DMAS_CELLS_BY_NAME(DT_DRV_INST(inst), cell_name, name)
+
+/**
+ * @brief Does a DMA client has a DMA specifer referenced as "name"
+ * @param node_id node identifier
+ * @param name dma channel specifier name
```
Comment (mbolivar-nordic) [inline include/devicetree/dma.h:216]: When you get a chance to update the docs, I usually put "lowercase-and-underscores" for any parameters that have to be converted that way.

**Review by mbolivar-nordic — COMMENTED** [review]
This looks sane to me

File: dts/bindings/i2s/st,stm32-i2s.yaml [hunk 6]
```diff
@@ -5,7 +5,7 @@ description: STM32 I2S controller
 
 compatible: "st,stm32-i2s"
 
-include: [i2s-controller.yaml, dma.yaml]
+include: [i2s-controller.yaml]
```
Comment (FRASTM) [inline dts/bindings/i2s/st,stm32-i2s.yaml:8]: we did include the dma bindings for the spi client, so why not for i2s

Comment (erwango) [reply dts/bindings/i2s/st,stm32-i2s.yaml:8]: dma client properties have been moved to base.yaml.
So they'll be available for all clients, no more need to include dma.yaml (which has been deleted)

File: scripts/dts/gen_defines.py (hunk 3 above)
Comment (erwango) [reply scripts/dts/gen_defines.py:493]: So do I need an additional update?

File: include/devicetree/dma.h (hunk 5 above)
Comment (erwango) [reply include/devicetree/dma.h]: Done

File: dts/bindings/base/base.yaml (hunk 4 above)
Comment (erwango) [reply dts/bindings/base/base.yaml]: fixed, txs

**Comment by erwango** [comment]
Thanks for the reviews. I'm closing and pushing https://github.com/zephyrproject-rtos/zephyr/pull/24204 as official PR

**Pull request closed without merging**